### DISCUSSION TITLE
UART功能在模擬時與上FPGA時不一致

### DISCUSSION THREAD

**Discussion by stage0831** [bol-edu/HLS-SOC-Discussions#160] [Q&A]
大家好，下面兩張圖是我在做模擬時的波型以及log，沒意外的話UART應該是有功能的，且中斷也是在main的最後才執行。

![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/72055234/4491cf3f-6713-40bf-8360-b523b6be1c7f)
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/72055234/4e771079-9cdb-4f43-a5dd-d987445914ee)

但是使用FPGA後，UART沒有傳回hello，如下圖。

![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/72055234/533e1bcf-50e2-4091-98cb-c9671ca4e664)

目前有點無從下手，想請問可能是哪個部分有出錯，或是能夠從哪邊下手debug ? 謝謝大家 !

**Comment by LingChiYang** [comment]
同學好

1. 我會先確認rx和interrupt是否可以正常運作，會暫時用mprj來代替tx來確認答案。
2. 接著獨立驗證tx是否可以正常運作(在main寫tx，不放在isr)
3. 若上述都正常，有可能是合成或block design有問題，再請同學附上log看問題在哪

**Reply by stage0831** [reply]
想請問助教，如附圖的波型，紅色圈起來的部份無法算是tx rx 和 interrupt正常工作的波型嗎?

![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/72055234/8f66c9ed-c8cd-47ae-b1e8-9d648c5aeb1d)

另外，在跑run_vivado時有跑出下圖的warning
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/72055234/691483c7-efbf-473a-b531-3397b1b2bbfb)

有看了一下report有兩條path有hold time violation，但我們有進到vivado並調整implementation的策略，並成功解決此問題，然而上FPGA也是沒有辦法傳回hello。

我們有嘗試過將有violation和沒有violation的都燒進板子，但結果都是沒有hello。目前的情況是這樣，謝謝助教的回覆!

**Reply by LingChiYang** [reply]
波型看起來沒問題，可以試試看上述的debug方法。
或是同學可以提供github repo讓我複製這個問題

**Reply by Wil0408** [reply]
大家好，我們這組也碰到了相同的問題，目前我debug的方式是把juputer notebook tx傳的值用mprj輸出，當我們傳不同的值時可以在jupyter notebook上觀察到mprj的變化。
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/87405718/90c3e421-9251-4619-ab7d-4be296b8dc3d)
於是我們又嘗試直接在uart.c的main function中直接呼叫isr.c中的uart_write()，並在jupyter notebook上確認有收到值，輸出一個字串。
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/87405718/8e278b4d-fe8f-4976-9ea7-4854f905c494)
想請問各位這有可能是什麼問題或後續我們應該如何debug呢?

**Reply by stage0831** [reply]
同學你好，我剛剛成功解決hello未出現的問題，我發現問題不是uart這邊，而是firmware的部分，如下圖

![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/72055234/fb0f5b58-bd7c-4dfd-a44a-eb06dbdca66c)

我嘗試了很久發現我上板子後，firmware部分會卡在*tmp = matmu()，因此若將中斷設定放在下面會執行不到

我將uart中斷的部分上移到函數上面就可以了，你試試看。

**Reply by stage0831** [reply]
但我目前的問題在於firmware卡在函數呼叫這邊，如下面兩張圖所示

![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/72055234/98346b39-51eb-4ed3-902b-f4332d732053)

![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/72055234/d36c292b-b42b-48f6-92df-9c9be333ff26)

若你有進展或是有想法在請你幫幫忙，謝謝!

**Reply by alexlu0414** [reply]
各位同學如果發現hello有問題，可以試試看這個神奇按鈕==
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/121178881/a321bd02-da52-46d5-8994-3b0fc1882bbd)

**Reply by Wil0408** [reply]
同學你好，我們最後的結論是如果有修改過hex檔jupyter notebook就要restart kernel，還是十分感謝你的回覆。
